src/FieldMapB.cpp

```diff
@@ -0,0 +1,377 @@
+        //std::cout << indices[0] << ", " << indices[1] << ", " << Bcomp[0] << ", " << Bcomp[1] << std::endl;
+      }
+    }
+    else {
+      coord.resize(3);
+      Bcomp.resize(3);
+      iss >> coord[0] >> coord[1] >> coord[2] >> Bcomp[0] >> Bcomp[1] >> Bcomp[2];
+
+      if( ! GetIndices(coord[0], coord[1], coord[2], &ix, &iy, &iz, &dx, &dy, &dz) ) {
+        std::cout << "FieldMapB Warning: coordinates out of range, skipped it."<< std::endl;
+      }
+      else {
+        Bvals_XYZ[ ix ][ iy ][ iz ] = { Bcomp[0] * scale, Bcomp[1] * scale, Bcomp[2] * scale };
+      }
+    }
+  }
+}
+
+// get field components
+void FieldMapB::fieldComponents(const double* pos, double* field)
+{
+  // coordinate conversion
+  auto p = trans_inv * ROOT::Math::XYZPoint(pos[0], pos[1], pos[2]);
+
+  if( case_BrBz ) {
+    // coordinates conversion
+    const double r   = sqrt(p.x() * p.x() + p.y() * p.y());
+    const double z   = p.z();
+    const double phi = atan2(p.y(), p.x());
+
+    if( ! GetIndices(r, z, &ir, &iz, &dr, &dz) ) {
+      // out of range
+      return;
+    }
+
+    // p1    p3
+    //    p
+    // p0    p2
+    auto& p0 = Bvals_RZ[ ir     ][ iz    ];
+    auto& p1 = Bvals_RZ[ ir     ][ iz + 1];
+    auto& p2 = Bvals_RZ[ ir + 1 ][ iz    ];
+    auto& p3 = Bvals_RZ[ ir + 1 ][ iz + 1];
+
+    // Bilinear interpolation
+    double Br = p0[0] * (1 - dr) * (1 - dz) + p1[0] * (1 - dr) * dz
+              + p2[0] *    dr    * (1 - dz) + p3[0] *    dr    * dz;
+
+    double Bz = p0[1] * (1 - dr) * (1 - dz) + p1[1] * (1 - dr) * dz
+              + p2[1] *    dr    * (1 - dz) + p3[1] *    dr    * dz;
+
+    // convert Br Bz to Bx By Bz
+    B_interpolated = ROOT::Math::XYZPoint(Br * sin(phi), Br * cos(phi), Bz);
+    B_interpolated = trans * B_interpolated;
+  }
+  else {
+
+    if( ! GetIndices(p.x(), p.y(), p.z(), &ix, &iy, &iz, &dx, &dy, &dz) ) {
+      return; // out of range
+    }
+
+    double b[3] = {0};
+    for(int comp = 0; comp < 3; comp++) { // field component loop
+      // Trilinear interpolation
+      // First along X, along 4 lines
+      double b00 = Bvals_XYZ[ ix     ][ iy     ][ iz     ][comp] * (1 - dx)
+                 + Bvals_XYZ[ ix  + 1][ iy     ][ iz     ][comp] * dx;
+      double b01 = Bvals_XYZ[ ix     ][ iy     ][ iz  + 1][comp] * (1 - dx)
+                 + Bvals_XYZ[ ix  + 1][ iy     ][ iz  + 1][comp] * dx;
+      double b10 = Bvals_XYZ[ ix     ][ iy  + 1][ iz     ][comp] * (1 - dx)
+                 + Bvals_XYZ[ ix  + 1][ iy  + 1][ iz     ][comp] * dx;
+      double b11 = Bvals_XYZ[ ix     ][ iy  + 1][ iz  + 1][comp] * (1 - dx)
+                 + Bvals_XYZ[ ix  + 1][ iy  + 1][ iz  + 1][comp] * dx;
+      // Next along Y, along 2 lines
+      double b0 = b00 * (1 - dy) + b10 * dy;
+      double b1 = b01 * (1 - dy) + b11 * dy;
+      // Finally along Z
+      b[comp] = b0 * (1 - dz) + b1 * dz;
+    }
+
+    B_interpolated = ROOT::Math::XYZPoint( b[0], b[1], b[2] );
+  }
+
+  field[0] += B_interpolated.x() * tesla;
```

Comment:
This unit conversion shouldn't be necessary to do on the interpolated value and should already have been done on reading in and storing values in the grid.

Reply:
Yes, we should multiply by tesla right when we read the file.

Reply:
In `LoadMap`

Reply:
Done.